include/gridtools/stencil_composition/backend_cuda/launch_kernel.hpp

```diff
@@ -0,0 +1,244 @@
+                    static constexpr auto boundary = ceil(-Extent::iminus::value);
+                    // we dedicate one warp to execute regions (a,h,e), so here we make sure we have enough threads
+                    GT_STATIC_ASSERT(jboundary_limit * boundary <= BlockSizeI, GT_INTERNAL_ERROR);
+
+                    i_block = -boundary + (int_t)threadIdx.x % boundary;
+                    j_block = (int_t)threadIdx.x / boundary + Extent::jminus::value;
+                } else {
+                    assert(Extent::iplus::value > 0);
+                    assert(threadIdx.y < iminus_limit + 1);
+                    static constexpr auto boundary = ceil(Extent::iplus::value);
+                    // we dedicate one warp to execute regions (c,i,g), so here we make sure we have enough threads
+                    GT_STATIC_ASSERT(jboundary_limit * boundary <= BlockSizeI, GT_INTERNAL_ERROR);
+
+                    i_block = (int_t)threadIdx.x % boundary + BlockSizeI;
+                    j_block = (int_t)threadIdx.x / boundary + Extent::jminus::value;
+                }
+
+                int_t i_block_size =
+                    (blockIdx.x + 1) * BlockSizeI < i_size ? BlockSizeI : i_size - blockIdx.x * BlockSizeI;
+                int_t j_block_size =
+                    (blockIdx.y + 1) * BlockSizeJ < j_size ? BlockSizeJ : j_size - blockIdx.y * BlockSizeJ;
+
+                fun(i_block, j_block, extent_validator_f<Extent>{i_block, j_block, i_block_size, j_block_size});
+            }
+
+            template <size_t NumThreads, int_t BlockSizeI, int_t BlockSizeJ, class Fun>
+            __global__ void __launch_bounds__(NumThreads)
+                zero_extent_wrapper(Fun const fun, int_t i_size, int_t j_size) {
+                if (blockIdx.x * BlockSizeI + threadIdx.x < i_size && blockIdx.y * BlockSizeJ + threadIdx.y < j_size)
+                    fun(threadIdx.x, threadIdx.y, dummy_validator_f());
+            }
+
+            template <size_t NumThreads, class Fun>
+            __global__ void __launch_bounds__(NumThreads) trivial_wrapper(Fun const fun) {
+                fun(threadIdx.x, threadIdx.y, dummy_validator_f());
+            }
+
+            template <class Extent>
+            constexpr bool is_empty_ij_extents() {
+                return Extent::iminus::value == 0 && Extent::iplus::value == 0 && Extent::jminus::value == 0 &&
+                       Extent::jplus::value == 0;
+            }
+
+            template <class Kernel, class... Args>
+            void launch(dim3 blocks, dim3 threads, size_t shared_memory_size, Kernel kernel, Args... args) {
+                GT_CUDA_CHECK(
+                    cudaFuncSetAttribute(kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, shared_memory_size));
+                kernel<<<blocks, threads, shared_memory_size>>>(std::move(args)...);
+#ifndef NDEBUG
+                GT_CUDA_CHECK(cudaDeviceSynchronize());
+#else
+                GT_CUDA_CHECK(cudaGetLastError());
```

Comment:
Maybe we should move cudaGetLastError() out of the conditional and also execute it in debug mode as we discovered that cudaDeviceSynchronize sometimes does not catch kernel launch failures in newer CUDA versions (or the V100).

Reply:
done